extensions/spatial/src/org/sbml/jsbml/validator/offline/constraints/GeometryConstraints.java

```diff
@@ -0,0 +1,420 @@
+		    case GENERAL_CONSISTENCY:
+		    	if(level >= 3){		    		
+		    		addRangeToSet(set, SPATIAL_23701, SPATIAL_23717);
+		    	}
+		      break;
+		    case IDENTIFIER_CONSISTENCY:
+		      break;
+		    case MATHML_CONSISTENCY:
+		      break;
+		    case MODELING_PRACTICE:
+		      break;
+		    case OVERDETERMINED_MODEL:
+		      break;
+		    case SBO_CONSISTENCY:
+		      break;
+		    case UNITS_CONSISTENCY:
+		      break;
+		    }
+	  }
+	  
+	  @Override
+	  public ValidationFunction<?> getValidationFunction(int errorCode, ValidationContext context){
+		  ValidationFunction<Geometry> func = null;
+		  
+		  switch (errorCode) {
+		  
+		  	case SPATIAL_23701:
+		  	{
+		  		// A Geometry object may have the optional SBML Level 3 Core attributes metaid and sboTerm. 
+		  		// No other attributes from the SBML Level 3 Core namespaces are permitted on a Geometry.
+		  		
+		  		func = new UnknownCoreAttributeValidationFunction<Geometry>();
+		  		break;
+		  	}
+		  	
+		  	case SPATIAL_23702:
+		  	{
+		  		// A Geometry object may have the optional SBML Level 3 Core subobjects for notes and annotations.
+		  		// No other elements from the SBML Level 3 Core namespaces are permitted on a Geometry.
+		  		
+		  		func = new UnknownCoreElementValidationFunction<Geometry>();
+		  		break;
+		  	}
+		  	
+		  	case SPATIAL_23703:
+		  	{
+		  		// A Geometry object must have the required attribute spatial:coordinateSystem, and may 
+		  		// have the optional attribute spatial:id. No other attributes from the SBML Level 3 Spatial
+		  		// Processes namespaces are permitted on a Geometry object.
+		  		
+		  		func = new UnknownPackageAttributeValidationFunction<Geometry>(SpatialConstants.shortLabel) {
+		  			
+		  			@Override
+		  			public boolean check(ValidationContext ctx, Geometry geom) {
+		  				
+		  				if(!geom.isSetCoordinateSystem()) {
+		  					return false;
+		  				}
+		  				
+		  				return super.check(ctx, geom);
+		  			}
+		  		};
+		  		break;
+		  	}
+		  	
+		  	case SPATIAL_23704:
+		  	{
+		  		// A Geometry object may contain one and only one instance of each of the ListOfCoordinateComponents,
+		  		// ListOfDomainTypes, ListOfDomains, ListOfAdjacentDomains, ListOfGeometryDefinitions and
+		  		// ListOfSampledFields elements. No other elements from the SBML Level 3 Spatial Processes
+		  		// namespaces are permitted on a Geometry object.
+		  		
+		  		func = new ValidationFunction<Geometry>() {
+		  			
+		  			@Override
+		  			public boolean check(ValidationContext ctx, Geometry geom) {
+		  				
+		  				boolean locc = geom.isSetListOfCoordinateComponents() 
+		  						&& new DuplicatedElementValidationFunction<Geometry>(SpatialConstants.listOfCoordinateComponents).check(ctx, geom);
+		  				boolean lodt = geom.isSetListOfDomainTypes() 
+		  						&& new DuplicatedElementValidationFunction<Geometry>(SpatialConstants.listOfDomainTypes).check(ctx, geom);
+		  				boolean lod = geom.isSetListOfDomains() 
```

Comment:
I don't think this logic is right. If a ListOf is not set, the boolean should be true and in your case, it is false.
You can create an additional test file where one or several of the listOf are not there at all and you will see that your current code will fail the constraint where it should not.

Reply:
@niko-rodrigue I just checked in the specification, and it says that listOfCoordinateComponents is always needed and rest are optional. So I think I'll retain the isSetListOf method for coordinate component and remove the method for rest of them?